WpfSendKeys/SendKeys.cs

```diff
@@ -1,120 +1,112 @@
-        {
-            PresentationSource presentationSource = PresentationSource.FromVisual(element);
-            int timestamp = Environment.TickCount;
-            KeyEventArgs args = new KeyEventArgs(keyboardDevice, presentationSource, timestamp, key);
-
-            // 1) PreviewKeyDown
-            args.RoutedEvent = Keyboard.PreviewKeyDownEvent;
-            element.RaiseEvent(args);
-
-            // 2) KeyDown
-            args.RoutedEvent = Keyboard.KeyDownEvent;
-            element.RaiseEvent(args);
-
-            // 3) TextInput
-            SendInputIfNecessary(element, key, modifiers, keyboardDevice);
-
-            // 4) PreviewKeyUp
-            args.RoutedEvent = Keyboard.PreviewKeyUpEvent;
-            element.RaiseEvent(args);
-
-            // 5) KeyUp
-            args.RoutedEvent = Keyboard.KeyUpEvent;
-            element.RaiseEvent(args);
-        }
-
-        private static void SendInputIfNecessary(UIElement element, Key key, ModifierKeys modifiers, KeyboardDevice keyboardDevice)
-        {
-            if (modifiers.HasFlag(ModifierKeys.Control) || modifiers.HasFlag(ModifierKeys.Alt))
-            {
-                return;
-            }
-
-            string input = "";
-
-            input = KeyboardLayout.Instance.GetInputForGesture(new KeyPressInfo(key, modifiers));
-            if (input == "")
-            {
-                input = GetInputFromKey(key);
-            }
-
-            if (string.IsNullOrEmpty(input))
-            {
-                return;
-            }
-
-            if (modifiers == ModifierKeys.Shift)
-            {
-                input = input.ToUpperInvariant();
-            }
-            else
-            {
-                input = input.ToLowerInvariant();
-            }
-
-            SendInput(element, input);
-        }
-
-        private static string GetInputFromKey(Key key)
-        {
-            switch (key)
-            {
-                case Key.Left:
-                case Key.Up:
-                case Key.Right:
-                case Key.Down:
-                case Key.Home:
-                case Key.End:
-                case Key.Insert:
-                case Key.Delete:
-                case Key.PageUp:
-                case Key.PageDown:
-                case Key.Back:
-                case Key.Escape:
-                case Key.Enter:
-                case Key.Tab:
-                case Key.Space:
-                    return "";
-            }
-            return new KeyConverter().ConvertToInvariantString(key);
-        }
-    }
-}
+	public class SendKeys
+	{
+		public static void Send(UIElement element, string text)
+		{
+			var sequence = Parse(text);
+			foreach (var keyPressInfo in sequence)
+			{
+				Send(element, keyPressInfo);
+			}
+		}
+
+		public static void Send(UIElement element, KeyPressInfo keyPressInfo)
+		{
+			KeyboardDevice keyboardDevice = InputManager.Current.PrimaryKeyboardDevice;
```

Comment:
I don't like it that we ask the OS for the current keyboard status.
I see that when I run the test and press SHIFT key, it shows up as a modifier. This may have unintended side effects.
We are mostly immune to them because `SendInputIfNecessary` looks at modifiers of `KeyPressInfo` and inserts correct character, but other Keyboard events pass in the incorrect modifier keys.
Can we use `MockKeyboardDevice` in all cases?